quinn-proto/src/connection/streams/state.rs

```diff
@@ -1553,4 +1581,77 @@ mod tests {
+        assert!(explanded);
+        assert_eq!(server.receive_window, new_receive_window as u64);
+        assert_eq!(server.local_max_data, new_receive_window as u64);
+        assert_eq!(server.receive_window_shrink_debt, 0);
+        let prev_local_max_data = server.local_max_data;
+
+        // credit, expecting all of them added to local_max_data
+        let credits = 1024u64;
+        let shuould_transmit = server.add_read_credits(credits);
+        assert_eq!(server.receive_window_shrink_debt, 0);
+        assert_eq!(server.local_max_data, prev_local_max_data + credits);
+        assert!(shuould_transmit.should_transmit());
+    }
+
+    #[test]
+    fn shrink_receive_window() {
+        let mut server = make(Side::Server);
+        let new_receive_window = 1024 * 512u32;
+        let prev_local_max_data = server.local_max_data;
+
+        // shrink the receive_winbow, local_max_data is not expected to be changed
+        let shrink_diff = server.receive_window - new_receive_window as u64;
+        let explanded = server.set_receive_window(new_receive_window.into());
+        assert!(!explanded);
+        assert_eq!(server.receive_window, new_receive_window as u64);
+        assert_eq!(server.local_max_data, prev_local_max_data);
+        assert_eq!(server.receive_window_shrink_debt, shrink_diff);
+        let prev_local_max_data = server.local_max_data;
+
+        // credit, local_max_data does not change as it is absorbed by receive_window_shrink_debt
+        let credits = 1024u64;
+        let expected_receive_window_shrink_debt = server.receive_window_shrink_debt - credits;
+        let shuould_transmit = server.add_read_credits(credits);
+        assert_eq!(
+            server.receive_window_shrink_debt,
+            expected_receive_window_shrink_debt
+        );
+        assert_eq!(server.local_max_data, prev_local_max_data);
+        assert!(!shuould_transmit.should_transmit());
+
+        // credit again, local_max_data does not change as it is absorbed by receive_window_shrink_debt
+        let expected_receive_window_shrink_debt = server.receive_window_shrink_debt - credits;
+        let shuould_transmit = server.add_read_credits(credits);
+        assert_eq!(
+            server.receive_window_shrink_debt,
+            expected_receive_window_shrink_debt
+        );
+        assert_eq!(server.local_max_data, prev_local_max_data);
+        assert!(!shuould_transmit.should_transmit());
+
+        // credit again which exceeds all remaining expected_receive_window_shrink_debt
+        let credits = 1024 * 512;
+        let expected_local_max_data =
+            server.local_max_data + (credits - server.receive_window_shrink_debt);
+        let shuould_transmit = server.add_read_credits(credits);
+        assert_eq!(server.receive_window_shrink_debt, 0);
+        assert_eq!(server.local_max_data, expected_local_max_data);
+        assert!(!shuould_transmit.should_transmit());
```

Comment:
If `credits - server.receive_window_shrink_debt` is expected to be nonzero:
- we should assert explicitly that `local_max_data` grew
- we should not assert that `should_transmit` is false to reduce this test's sensitivity to the flow control significance threshold

Reply:
Done